### ISSUE TITLE
SegFault in tiny_dnn::core::connection_table::is_connected

### ISSUE THREAD

**Issue by Sagre** [tiny-dnn/tiny-dnn#962]
I obtained a segfault after testing a trained network after the first epoch. I started with the mnist example, added the image convertion from https://github.com/tiny-dnn/tiny-dnn/wiki/Data-Format for my own image collection and played around with the network design. The last thing i changed, was partly switching from tanh_layers to leaky_relu_layers, so maybe it is caused by that.

Backtrace:
`#0  0x000000000072f5e3 in tiny_dnn::core::connection_table::is_connected(unsigned long, unsigned long) const (this=0x7fff3a7f0620, x=0, y=32) at /home/sagre/caros/tiny-dnn/tiny_dnn/core/params/conv_params.h:53
#1  0x000000000073a458 in tiny_dnn::kernels::conv2d_op_internal(std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const&, std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > const&, std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > const&, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >&, tiny_dnn::core::conv_params const&, bool)::{lambda(tiny_dnn::blocked_range const&)#1}::operator()(tiny_dnn::blocked_range const&) const (__closure=0x7fff3a7f04f0, r=...) at /home/sagre/caros/tiny-dnn/tiny_dnn/core/kernels/conv2d_op_internal.h:37
#2  0x000000000075d95f in tiny_dnn::xparallel_for<tiny_dnn::kernels::conv2d_op_internal(std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const&, std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > const&, std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > const&, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >&, tiny_dnn::core::conv_params const&, bool)::{lambda(tiny_dnn::blocked_range const&)#1}>(unsigned long, unsigned long, tiny_dnn::kernels::conv2d_op_internal(std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const&, std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > const&, std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > const&, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >&, tiny_dnn::core::conv_params const&, bool)::{lambda(tiny_dnn::blocked_range const&)#1} const&) (begin=0, end=60, f=...) at /home/sagre/caros/tiny-dnn/tiny_dnn/util/parallel_for.h:79
#3  0x000000000075d91e in tiny_dnn::parallel_for<tiny_dnn::kernels::conv2d_op_internal(std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const&, std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > const&, std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > const&, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >&, tiny_dnn::core::conv_params const&, bool)::{lambda(tiny_dnn::blocked_range const&)#1}>(unsigned long, unsigned long, tiny_dnn::kernels::conv2d_op_internal(std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const&, std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > const&, std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > const&, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >&, tiny_dnn::core::conv_params const&, bool)::{lambda(tiny_dnn::blocked_range const&)#1} const&, unsigned long) (begin=0, end=60, f=...) at /home/sagre/caros/tiny-dnn/tiny_dnn/util/parallel_for.h:129
#4  0x000000000074e183 in tiny_dnn::for_<unsigned long, tiny_dnn::kernels::conv2d_op_internal(std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const&, std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > const&, std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > const&, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >&, tiny_dnn::core::conv_params const&, bool)::{lambda(tiny_dnn::blocked_range const&)#1}>(bool, unsigned long, unsigned long, tiny_dnn::kernels::conv2d_op_internal(std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const&, std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > const&, std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > const&, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >&, tiny_dnn::core::conv_params const&, bool)::{lambda(tiny_dnn::blocked_range const&)#1}, unsigned long) (parallelize=true, begin=0, end=60, f=..., grainsize=0) at /home/sagre/caros/tiny-dnn/tiny_dnn/util/parallel_for.h:179
#5  0x000000000073a773 in tiny_dnn::kernels::conv2d_op_internal(std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const&, std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > const&, std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > const&, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >&, tiny_dnn::core::conv_params const&, bool) (in_data=std::vector of length 60, capacity 60 = {...}, W=std::vector of length 165888, capacity 165888 = {...}, bias=std::vector of length 32, capacity 32 = {...}, out_data=std::vector of length 60, capacity 60 = {...}, params=..., parallelize=true)
    at /home/sagre/caros/tiny-dnn/tiny_dnn/core/kernels/conv2d_op_internal.h:19
#6  0x000000000073ac84 in tiny_dnn::kernels::conv2d_op_avx(std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const&, std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > const&, std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > const&, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >&, tiny_dnn::core::conv_params const&, bool) (in_data=std::vector of length 60, capacity 60 = {...}, W=std::vector of length 165888, capacity 165888 = {...}, bias=std::vector of length 32, capacity 32 = {...}, out_data=std::vector of length 60, capacity 60 = {...}, params=..., layer_parallelize=true)
    at /home/sagre/caros/tiny-dnn/tiny_dnn/core/kernels/conv2d_op_avx.h:467
#7  0x000000000073af78 in tiny_dnn::Conv2dOp::compute(tiny_dnn::core::OpKernelContext&) (this=0x552c9f0, context=...) at /home/sagre/caros/tiny-dnn/tiny_dnn/core/kernels/conv2d_op.h:46
#8  0x000000000073bed4 in tiny_dnn::convolutional_layer::forward_propagation(std::vector<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >*, std::allocator<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >*> > const&, std::vector<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >*, std::allocator<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >*> >&) (this=0x552c390, in_data=std::vector of length 3, capacity 3 = {...}, out_data=std::vector of length 1, capacity 1 = {...}) at /home/sagre/caros/tiny-dnn/tiny_dnn/layers/convolutional_layer.h:271
#9  0x0000000000732955 in tiny_dnn::layer::forward() (this=0x552c390) at /home/sagre/caros/tiny-dnn/tiny_dnn/layers/layer.h:554
#10 0x0000000000734fca in tiny_dnn::sequential::forward(std::vector<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >, std::allocator<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > > > const&) (this=0x7fff3a7f0e60, first=std::vector of length 60, capacity 60 = {...}) at /home/sagre/caros/tiny-dnn/tiny_dnn/nodes.h:296
#11 0x00000000007799bb in tiny_dnn::network<tiny_dnn::sequential>::fprop(std::vector<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >, std::allocator<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > > > const&) (this=0x7fff3a7f0e40, in=std::vector of length 60, capacity 60 = {...}) at /home/sagre/caros/tiny-dnn/tiny_dnn/network.h:203
#12 0x000000000077b479 in tiny_dnn::network<tiny_dnn::sequential>::train_onebatch<tiny_dnn::mse, tiny_dnn::adagrad>(tiny_dnn::adagrad&, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const*, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::ali---Type <return> to continue, or q <return> to quit---
gned_allocator<float, 64ul> > > > const*, int, int, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const*) (this=0x7fff3a7f0e40, optimizer=..., in=0xdeddaf0, t=0xe875b30, batch_size=60, num_tasks=8, t_cost=0x0) at /home/sagre/caros/tiny-dnn/tiny_dnn/network.h:924
#13 0x0000000000772218 in tiny_dnn::network<tiny_dnn::sequential>::train_once<tiny_dnn::mse, tiny_dnn::adagrad>(tiny_dnn::adagrad&, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const*, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const*, int, int, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > const*) (this=0x7fff3a7f0e40, optimizer=..., in=0xdeddaf0, t=0xe875b30, size=60, nbThreads=8, t_cost=0x0) at /home/sagre/caros/tiny-dnn/tiny_dnn/network.h:897
#14 0x0000000000729bb6 in tiny_dnn::network<tiny_dnn::sequential>::fit<tiny_dnn::mse, tiny_dnn::adagrad, train_lenet(const string&, const string&, const string&)::<lambda()>, train_lenet(const string&, const string&, const string&)::<lambda()> >(tiny_dnn::adagrad &, const std::vector<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64> > > >, std::allocator<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64> > > > > > &, const std::vector<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64> > > >, std::allocator<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64> > > > > > &, size_t, int, <lambda()>, <lambda()>, bool, int, const std::vector<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64> > > >, std::allocator<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64> > > > > > &) (this=0x7fff3a7f0e40, optimizer=..., inputs=std::vector of length 1819, capacity 1819 = {...}, desired_outputs=std::vector of length 1819, capacity 1819 = {...}, batch_size=60, epoch=5, on_batch_enumerate=..., on_epoch_enumerate=..., reset_weights=false, n_threads=8, t_cost=std::vector of length 0, capacity 0)
    at /home/sagre/caros/tiny-dnn/tiny_dnn/network.h:863
#15 0x00000000007293ef in tiny_dnn::network<tiny_dnn::sequential>::train<tiny_dnn::mse, tiny_dnn::adagrad, train_lenet(const string&, const string&, const string&)::<lambda()>, train_lenet(const string&, const string&, const string&)::<lambda()> >(tiny_dnn::adagrad &, const std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64> > > > &, const std::vector<unsigned long, std::allocator<unsigned long> > &, size_t, int, <lambda()>, <lambda()>, bool, int, const std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64> > > > &) (this=0x7fff3a7f0e40, optimizer=..., inputs=std::vector of length 1819, capacity 2048 = {...}, class_labels=std::vector of length 1819, capacity 2048 = {...}, batch_size=60, epoch=5, on_batch_enumerate=..., on_epoch_enumerate=..., reset_weights=false, n_threads=8, t_cost=std::vector of length 0, capacity 0)
`

Network design:

nn << convolutional_layer(128, 128, 5, 1, 64, 
                              padding::valid, true, 1, 1, backend_type)
        << leaky_relu_layer()
        << average_pooling_layer(124,124, 64, 2) 
        << leaky_relu_layer()
        << convolutional_layer(62, 62, 9, 64, 32, 
                              connection_table(tbl, 6, 16), padding::valid, true,
                              1, 1, backend_type)
       << leaky_relu_layer()
        << average_pooling_layer(54, 54, 32, 2) 
       << tanh_layer(27, 27, 32)
       << convolutional_layer(27, 27, 5, 32,60,  
                              padding::valid, true, 1, 1, backend_type)
       << tanh_layer(23,23,60)
       << fully_connected_layer(31740, 10, true,backend_type)
       << tanh_layer(10);

Train function:

`
static void train_lenet(const std::string& data_dir_path, const std::string& test_dir_path, const std::string& output_path) {

    
    network<sequential> nn;
    adagrad optimizer;

    construct_net(nn);

    std::cout << "load models..." << std::endl;

    // load MNIST dataset
    std::vector<label_t> train_labels;
    std::vector<vec_t> train_images;
    std::vector<label_t> test_labels;
    std::vector<vec_t> test_images;

    convert_images(data_dir_path,1.0,128,128,train_images,train_labels);
    convert_images(test_dir_path,1.0,128,128,test_images,test_labels);

    std::cout << "start training, images: " << train_images.size() << ", " << train_labels.size() << std::endl;

    progress_display disp(static_cast<unsigned long>(train_images.size()));
    timer t;
    int minibatch_size = 60;
    int num_epochs = 5;

    optimizer.alpha *= static_cast<tiny_dnn::float_t>(0.01);
    auto on_enumerate_epoch = [&](){
        std::cout << t.elapsed() << "s elapsed." << std::endl;
        tiny_dnn::result res = nn.test(test_images, test_labels);
        std::cout << res.num_success << "/" << res.num_total << std::endl;
        
	disp.restart(static_cast<unsigned long>(train_images.size()));
        t.restart();
    };

    auto on_enumerate_minibatch = [&](){
        disp += minibatch_size;
    };

    nn.train<mse>(optimizer, train_images, train_labels, minibatch_size, num_epochs,
             on_enumerate_minibatch, on_enumerate_epoch);

    std::cout << "end training." << std::endl;

`
